File: ctci/java/ch1-arrays-and-strings/q1_is_unique/QuestionB.java
package q1_is_unique;

/**
 * Created by jjlee on 7/2/16.
 */
public class QuestionB {
    /* use a bit vector and assume lowercase alphabets (i.e. 26 characters) */
    public static boolean isUniqueNoAdditionalDataStructure(String s) {
        if (s.length() > 26)
            return false;

        int exists = 0;
        for (int i = 0; i < s.length(); i++) {
            int bitMask = 1 << (s.charAt(i) - 'a');
            if ((exists & bitMask) > 0)
                return false;
            else
                exists |= bitMask;
        }
        return true;
    }
}

File: ctci/python/ch1-arrays-and-strings/q1-is-unique/QuestionB.py
# use a bit vector and assume lowercase alphabets (i.e. 26 characters)
def is_unique_no_additional_data_structure(s):
    if len(s) > 26:
        return False

    exists = 0
    ord_a = ord('a')
    for c in s:
        bit_mask = 1 << (ord(c) - ord_a)
        if (exists & bit_mask) > 0:
            return False
        exists |= bit_mask
    return True
